### PR TITLE
GQSP template and angle solver

### PR DESCRIPTION
**Pull request by KetpuntoG**
- [x] Add thumbnail
- [x] Add changelog

This PR introduces the [Generalized QSP](https://arxiv.org/abs/2308.01501) template and some functionality to calculate the desire polynomials.

[sc-76734]

### CODE REVIEW

File: pennylane/templates/subroutines/gqsp.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        angles (array[float]): An array of angles with shape `(3, d+1)`, where `d` is the degree of the desired polynomial.
+            These angles define the coefficients for the polynomial transformation applied to the unitary matrix.
+
+        control (Union[Wires, int, str]): The control qubit used to encode the polynomial transformation on the unitary
+            operator.
+
+    .. note::
+
+        To calculate the angles given a polynomial ``poly``, it can be used the
+        function :func:`~.math.poly_to_angles` by choosing the routine ``"GQSP"``.
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:56]:
```suggestion
        :func:`~.math.poly_to_angles` can be used to calculate the angles given a polynomial ``poly`` with ``routine="GQSP"``.
```

File: pennylane/templates/subroutines/gqsp.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    Example:
+
+    .. code-block:: python
+
+        # P(x) = -0.1 + 0.2j x + 0.3 x^2
+        poly = [0.1, 0.2j, 0.3]
+
+        angles = qml.math.poly_to_angles(poly, "GQSP")
+
+        @qml.prod # transforms the qfunc into an Operator
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:67]: Nice! 🧑‍🍳

File: pennylane/templates/subroutines/gqsp.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        def unitary(wires):
+            qml.RX(0.3, wires)
+
+        dev = qml.device("default.qubit")
+
+        @qml.qnode(dev)
+        def circuit(angles):
+            qml.GQSP(unitary(wires = 1), angles, control = 0)
+            return qml.state()
+
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:77]:
```suggestion
```

File: pennylane/templates/subroutines/gqsp.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        matrix = qml.matrix(circuit, wire_order=[0, 1])(angles)
+
+    .. code-block:: pycon
+
+        >>> print(np.round(matrix,3)[:2, :2])
+        [[0.387+0.198j 0.03 -0.089j]
+        [0.03 -0.089j 0.387+0.198j]]
+
+
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:87]:
```suggestion
```

File: pennylane/templates/subroutines/gqsp.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        op_list = []
+
+        # These four gates adapt PennyLane's qml.U3 to the chosen U3 format in the GQSP paper.
+        op_list.append(qml.X(control))
+        op_list.append(qml.U3(2 * thetas[0], phis[0], lambds[0], wires=control))
+        op_list.append(qml.X(control))
+        op_list.append(qml.Z(control))
+
+        for theta, phi, lam in zip(thetas[1:], phis[1:], lambds[1:]):
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:159]: Prefer using `lamb` instead of `lam`

File: pennylane/templates/subroutines/gqsp.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+from pennylane.queuing import QueuingManager
+from pennylane.wires import Wires
+
+
+class GQSP(Operation):
+    r"""
+    Implements the generalized quantum signal processing `(GQSP) <https://arxiv.org/abs/2308.01501>`__ circuit.
+
+    This algorithm takes as input a unitary matrix in the form of an operator and encodes, with the help of an auxiliary
+    qubit, a polynomial transformation of the unitary.
```
Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:32]:
```suggestion
    qubit, a polynomial transformation of the unitary as shown below:
```

Comment (obliviateandsurrender) [inline pennylane/templates/subroutines/gqsp.py:32]:
```suggestion
    qubit, a polynomial transformation of the unitary as shown below:
```

**Review by obliviateandsurrender — COMMENTED** [review]
Removing my request changes, so not to be a blocker.

**Review by obliviateandsurrender — APPROVED** [review]

File: pennylane/templates/subroutines/gqsp.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:1]:
```suggestion
# Copyright 2024 Xanadu Quantum Technologies Inc.
```

File: pennylane/templates/subroutines/gqsp.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane.queuing import QueuingManager
+from pennylane.wires import Wires
+
+
+class GQSP(Operation):
+    r"""
+    Implements the generalized quantum signal processing `(GQSP) <https://arxiv.org/abs/2308.01501>`__ circuit.
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:29]:
```suggestion
    Implements the generalized quantum signal processing (GQSP) circuit.
```

File: pennylane/templates/subroutines/gqsp.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+    This algorithm takes as input a unitary matrix in the form of an operator and encodes, with the help of an auxiliary
+    qubit, a polynomial transformation of the unitary as shown below:
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:32]:
```suggestion
    This operation encodes a polynomial transformation of an input unitary operator following the algorithm 
    described in `arXiv:2308.01501 <https://arxiv.org/abs/2308.01501>`__ as:
```

File: pennylane/templates/subroutines/gqsp.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+    .. math::
+         U
+         \xrightarrow{GQSP}
+         \begin{pmatrix}
+         \text{poly}(U) & * \\
+         * & * \\
+         \end{pmatrix}
+
+
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:42]:
```suggestion
    The implementation requires one auxiliary qubit.
```

File: pennylane/templates/subroutines/gqsp.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    Args:
+
+        unitary (Operator): The unitary operator to be applied in the generalized quantum signal processing (GQSP) circuit.
+        angles (tensor[float]): An array of angles with shape `(3, d+1)`, where `d` is the degree of the desired polynomial.
+            These angles define the coefficients for the polynomial transformation applied to the unitary matrix.
+        control (Union[Wires, int, str]): The control qubit used to encode the polynomial transformation on the unitary
+            operator.
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:49]:
```suggestion
        unitary (Operator): the operator to be encoded by the GQSP circuit
        angles (tensor[float]): array of angles defining the polynomial transformation. The shape of the array must be `(3, d+1)`, where `d` is the degree of the polynomial.
        control (Union[Wires, int, str]): control qubit used to encode the polynomial transformation
```

File: pennylane/templates/subroutines/gqsp.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+         \xrightarrow{GQSP}
+         \begin{pmatrix}
+         \text{poly}(U) & * \\
+         * & * \\
+         \end{pmatrix}
+
+
+    Args:
+
+        unitary (Operator): The unitary operator to be applied in the generalized quantum signal processing (GQSP) circuit.
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:45]: Is the name "unitary" consistent with the PRD?

File: pennylane/templates/subroutines/gqsp.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        angles (tensor[float]): An array of angles with shape `(3, d+1)`, where `d` is the degree of the desired polynomial.
+            These angles define the coefficients for the polynomial transformation applied to the unitary matrix.
+        control (Union[Wires, int, str]): The control qubit used to encode the polynomial transformation on the unitary
+            operator.
+
+    .. note::
+
+        :func:`~.poly_to_angles` can be used to calculate the angles given a polynomial ``poly`` with ``routine="GQSP"``.
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:53]:
```suggestion
       The  :func:`~.poly_to_angles` function can be used to calculate the angles for a given polynomial.
```

File: pennylane/templates/subroutines/gqsp.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+    Example:
+
+    .. code-block:: python
+
+        # P(x) = -0.1 + 0.2j x + 0.3 x^2
+        poly = [0.1, 0.2j, 0.3]
+
+        angles = qml.math.poly_to_angles(poly, "GQSP")
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:62]:
```suggestion
        angles = qml.poly_to_angles(poly, "GQSP")
```

File: pennylane/templates/subroutines/qsvt.py [hunk 15]
```diff
@@ -571,6 +571,65 @@ def _compute_qsp_angle(poly_coeffs):
...
+        poly_coeffs (tensor-like): Coefficients of the input polynomial P.
+
+    Returns:
+        (tensor-like): QSP angles corresponding to the input polynomial P. The shape is (3, P-degree)
+    """
+
+    complementary = _complementary_poly(poly_coeffs)
+
+    def gqsp_u3_gate(theta, phi, lambd):
+        # Matrix definition of U3 gate chosen in the GQSP paper
```
Comment (soranjh) [inline pennylane/templates/subroutines/qsvt.py:591]: Please site the paper properly. The future developer will be confused about what "GQSP paper" means.

Also, it should be a private function and be outside of this function, unless something blocks that.

File: pennylane/templates/subroutines/qsvt.py [hunk 16]
```diff
@@ -571,6 +571,65 @@ def _compute_qsp_angle(poly_coeffs):
...
+            [
+                [exp_lambda_phi * qml.math.cos(theta), exp_phi * qml.math.sin(theta)],
+                [exp_lambda * qml.math.sin(theta), -qml.math.cos(theta)],
+            ],
+            dtype=complex,
+        )
+
+        return matrix
+
+    # Algorithm 1 in [arXiv:2308.01501]
```
Comment (soranjh) [inline pennylane/templates/subroutines/qsvt.py:606]:
```suggestion
    # From Algorithm 1 in [arXiv:2308.01501]
```

File: tests/templates/test_subroutines/test_gqsp.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_gqsp.py:1]:
```suggestion
# Copyright 2024 Xanadu Quantum Technologies Inc.
```

File: tests/templates/test_subroutines/test_gqsp.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    @pytest.mark.parametrize(
+        ("unitary", "poly"),
+        [
+            (qml.RX(0.3, wires=1), [0.3, -0.2j, 0.1]),
+            (qml.RZ(1.3, wires=1), [-0.3j, -0.2j, 0, 0, 0.2]),
+            (qml.RY(0.2, wires=1), [0.4, -0, 0.2, 0, 0.2]),
+        ],
+    )
+    def test_correct_solution(self, unitary, poly):
+        """Test that poly_to_angles and GQSP produce the correct solution"""
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_gqsp.py:52]: [optional] better to have another separate test that only checks GQSP accuracy with a set of given angles.

The current test, does two separate tests at the same time: GQSP and  poly_to_angles.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @KetpuntoG, addes some small comments. Looks good overall.

File: tests/templates/test_subroutines/test_gqsp.py (hunk 18 above)
Comment (soranjh) [reply tests/templates/test_subroutines/test_gqsp.py:52]: You can leave this test. My suggestion is to add an extra test that does not call poly_to_angles, instead uses angles computed with it (or another reference).

File: pennylane/templates/subroutines/gqsp.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+    .. note::
+
+       The  :func:`~.poly_to_angles` function can be used to calculate the angles for a given polynomial.
+
+    Example:
+
+    .. code-block:: python
+
+        # P(x) = -0.1 + 0.2j x + 0.3 x^2
```
Comment (soranjh) [inline pennylane/templates/subroutines/gqsp.py:58]:
```suggestion
        # P(x) = 0.1 + 0.2j x + 0.3 x^2
```

**Review by soranjh — APPROVED** [review]
Good to go :shipit:

**Pull request merged**